### PR TITLE
✨ Add support for storyNextUp, to allow default auto-advance on last page.

### PR DESCRIPTION
**Pull request by zaparent**
* Add maybeSetStoryNextUp_() to set default auto-advance-after value.
* Set `auto-advance-after` attribute to the value of `storyNextUp` param if (a) the current page is the last page, (b) the current page has no `auto-advance-after` set by the publisher AND (c) `storyNextUp` is set.
* Add isLastPage().

### CODE REVIEW

File: extensions/amp-story/1.0/amp-story-page.js [hunk 1]
```diff
@@ -378,6 +382,24 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+  /** @private */
+  maybeSetStoryNextUp_() {
+    const autoAdvanceAttr = this.element.getAttribute('auto-advance-after');
+    const storyNextUpParam = Services.viewerForDoc(this.element).getParam(
+      'storyNextUp'
+    );
+    if (
+      this.isLastPage() &&
+      autoAdvanceAttr === null &&
+      storyNextUpParam !== null
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:394]: Nit: please move the `storyNextUpParam` as the first condition so we save one `querySelector` when possible

File: extensions/amp-story/1.0/amp-story-page.js [hunk 2]
```diff
@@ -397,6 +419,19 @@ export class AmpStoryPage extends AMP.BaseElement {
     return this.isPrerenderActivePage_;
   }
 
+  /**
+   * Returns true if the page is the last amp-story-page in the amp-story.
+   * @return {boolean}
+   */
+  isLastPage() {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:426]: Nit: `@private` and a trailing underscore `isLastPage_`

File: extensions/amp-story/1.0/amp-story-page.js [hunk 3]
```diff
@@ -397,6 +419,19 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+    if (this.isLastPage_ != null) {
+      return this.isLastPage_;
+    }
+    const selector = 'amp-story-page:last-of-type';
+    const selectorNode = this.win.document.querySelector(selector);
+    return selectorNode == this.element;
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:432]: Maybe we can make this a tiny bit easier by using the `matches` helper:

https://github.com/ampproject/amphtml/blob/ea2e7f696feca04e4fc0be41f4dc60c9a240904d/src/dom.js#L483-L500

**Review by gmajoulet — COMMENTED** [review]
Awesome!! Just a few nits :))

File: extensions/amp-story/1.0/amp-story-page.js (hunk 1 above)
Comment (zaparent) [reply extensions/amp-story/1.0/amp-story-page.js]: Sounds good, how does this look?

File: extensions/amp-story/1.0/amp-story-page.js (hunk 2 above)
Comment (zaparent) [reply extensions/amp-story/1.0/amp-story-page.js]: Done. I had to change the name of the private variable (and I wasn't using it correctly for caching the value anyway).

File: extensions/amp-story/1.0/amp-story-page.js (hunk 3 above)
Comment (zaparent) [reply extensions/amp-story/1.0/amp-story-page.js]: Done. Love it, this is great!

File: extensions/amp-story/1.0/amp-story-page.js [hunk 4]
```diff
@@ -378,6 +383,23 @@ export class AmpStoryPage extends AMP.BaseElement {
...
+  /** @private */
+  maybeSetStoryNextUp_() {
+    const autoAdvanceAttr = this.element.getAttribute('auto-advance-after');
+    if (!this.isLastPage_() || autoAdvanceAttr !== null) {
+      return;
+    }
+    const storyNextUpParam = Services.viewerForDoc(this.element).getParam(
+      'storyNextUp'
+    );
+    if (storyNextUpParam !== null) {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story-page.js:395]: Sorry, it's super nitty but what I meant in my first comment is that we should check for this param first, before we do the `matches()` selector that don't need to be performed in most cases. JS <> DOM operations are usually what slows down the code so a good rule of thumb is to run as little as possible of these, but it's super nitty and doesn't really make any diff.

**Review by gmajoulet — APPROVED** [review]
LGTM :))

File: extensions/amp-story/1.0/amp-story-page.js (hunk 4 above)
Comment (zaparent) [reply extensions/amp-story/1.0/amp-story-page.js]: Ahhhh, I see. You made sense, I jsut misinterpreted. But I think we've got it now!

File: extensions/amp-story/1.0/amp-story-page.js [hunk 5]
```diff
@@ -378,6 +383,24 @@ export class AmpStoryPage extends AMP.BaseElement {
...
 
+  /** @private */
+  maybeSetStoryNextUp_() {
+    const autoAdvanceAttr = this.element.getAttribute('auto-advance-after');
+    const storyNextUpParam = Services.viewerForDoc(this.element).getParam(
+      'storyNextUp'
+    );
+    if (
+      storyNextUpParam !== null &&
+      autoAdvanceAttr === null &&
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-page.js:394]: nit: switch these so they read in the same order as the const declarations above.

File: extensions/amp-story/1.0/amp-story-page.js [hunk 6]
```diff
@@ -378,6 +383,24 @@ export class AmpStoryPage extends AMP.BaseElement {
     }
   }
 
+  /** @private */
+  maybeSetStoryNextUp_() {
```
Comment (processprocess) [inline extensions/amp-story/1.0/amp-story-page.js:387]: nit: A comment here to give context on storyNextUp could be helpful for scanning the code.

**Comment by gmajoulet** [comment]
/to @processprocess to merge it when ready :))

File: extensions/amp-story/1.0/amp-story-page.js (hunk 5 above)
Comment (zaparent) [reply extensions/amp-story/1.0/amp-story-page.js]: Done. Good call.

File: extensions/amp-story/1.0/amp-story-page.js (hunk 6 above)
Comment (zaparent) [reply extensions/amp-story/1.0/amp-story-page.js]: Done. Let me know if this looks good to you.

**Review by processprocess — APPROVED** [review]

**Pull request merged**